tests/integration/lxd/test_launcher.py

```diff
@@ -164,13 +166,84 @@ def test_launch_use_base_instance(get_base_instance, launch_args, instance_name)
         )
 
-        assert instance.exists()
-        assert instance.is_running()
-
         # confirm the same instance was launched with both fingerprints
         instance.execute_run(["stat", "/base-instance"], check=True)
         instance.execute_run(["stat", "/instance"], check=True)
 
+        assert instance.exists()
+        assert instance.is_running()
+    finally:
+        if instance.exists():
+            instance.delete()
+        if base_instance.exists():
+            base_instance.delete()
+
+
+@freeze_time(datetime.now() + timedelta(days=91))
+def test_launch_use_base_instance_expired(get_base_instance, instance_name):
+    """Launch an instance using an expired base instance.
+
+    First, launch an instance from an image and create a base instance.
+    Then launch an instance from the expired base instance, which will trigger the
+    creation of a new instance and base instance.
+
+    The LXD instance is created via subprocess, so the creation date the instance is
+    out of freezegun's scope and can't be modified.
+    """
+    base_instance = get_base_instance()
+    base_configuration = bases.BuilddBase(alias=bases.BuilddBaseAlias.FOCAL)
+
+    instance = lxd.launch(
+        name=instance_name,
+        base_configuration=base_configuration,
+        image_name="20.04",
+        image_remote="ubuntu",
+        use_base_instance=True,
+    )
+
+    try:
+        # the instance and base instance should both exist
+        assert instance.exists()
+        assert base_instance.exists()
+
+        # only the instance should be running
+        assert instance.is_running()
+        assert not base_instance.is_running()
+
+        # fingerprint the expired base instance
+        base_instance.start()
+        base_instance.execute_run(["touch", "/base-instance"])
+        base_instance.stop()
+
+        # delete the instance so a new instance is created from the base instance
+        instance.delete()
+        instance = lxd.launch(
+            name=instance_name,
+            base_configuration=base_configuration,
+            image_name="20.04",
+            image_remote="ubuntu",
+            use_base_instance=True,
+        )
+
+        assert instance.exists()
+        assert instance.is_running()
+
+        # confirm instance does not have the expired base instance's fingerprint
+        proc = instance.execute_run(
+            ["stat", "/base-instance"], capture_output=True, text=True
+        )
+        assert proc.returncode == 1
+        assert "'/base-instance': No such file or directory" in proc.stderr
+
+        # confirm new base instance does not have the expired base instance's
+        # fingerprint
+        base_instance.start()
+        proc = base_instance.execute_run(
+            ["stat", "/base-instance"], capture_output=True, text=True
+        )
+        assert proc.returncode == 1
+        assert "'/base-instance': No such file or directory" in proc.stderr
```

Comment:
It feels like there are multiple tests in here. Can this be split into a few different, more specific tests?

Whether it can or not, this might also be a good use case for [pytest-check](https://github.com/okken/pytest-check), so we can complete the entire test and provide all the failures in one go (or groups of failures using the `any_failures()` function).

Reply:
The purpose of this test is to follow this specific flow:
![image](https://user-images.githubusercontent.com/60674096/207079678-8348e569-2aa3-4d38-ad82-ee7032a62835.png)

To setup for this, I first need to run through this flow:
![image](https://user-images.githubusercontent.com/60674096/207081308-54f04ae4-5a20-4f8d-a439-8d926435f974.png)

This setup flow is tested in it's own separate test as well.  I added `asserts` to ensure the setup is successful.  Perhaps it could be moved to a fixture?  It's used by `test_launch_use_base_instance` and `test_launch_use_base_instance_expired`, so it would save some duplication and make the purpose of the test more clear.

Regarding the idea for pytest-check, I'm not sure. The function under test is sequential enough that once a single assert fails I won't benefit from seeing the subsequent failures.  And it may be more of a hindrance since bad calls to LXD can hang for a long time.

Reply:
That's the same image twice, but it was enough for me to me comfortable with what you're doing and agree with your implementation.

I've found a couple of other places where we should be doing something that runs multiple asserts without failing fast, but it does sound this is very much not the place.

I'm happy with this.

Reply:
Unfortunately you'll still need to get someone else on the team to approve since this newbie doesn't have write access yet :stuck_out_tongue:

Reply:
Thanks!

Although those images are slightly different - I showed the path through by bolding the boxes and arrows in the flowchart.  The first image goes through two more steps in the middle of the flowchart than the second image.

Reply:
Ohhhh I see now! The bolding didn't show up clearly on my phone. Sorry about that!